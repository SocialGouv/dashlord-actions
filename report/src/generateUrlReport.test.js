// @ts-nocheck

const fs = require("fs");
const generateUrlReport = require("./generateUrlReport");

jest.mock("fs");

const testUrl = "https://www.test.com";
const b64 = "aHR0cHM6Ly93d3cudGVzdC5jb20=";

// mock a json in the latest scan
const mockJson = (name, data) =>
  jest.mock(
    `results/${b64}/${name}`,
    () =>
      data
        ? data
        : {
          data: name,
        },
    { virtual: true }
  );

// for some reason jest.resetModules doesnt work with virtual JSON mocks
const unMockJson = (name) => {
  jest.unmock(`results/${b64}/${name}`);
};

describe("generateUrlReport", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test(`should return null on invalid url`, () => {
    expect(
      generateUrlReport({
        url: "https://www.invalid.com",
      })
    ).toEqual(null);
  });

  test(`should generate latest report for a valid url`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => false); // screenshot

    // mock all required files
    mockJson("codescanalerts.json", {
      report: "codescanalerts.json",
      totalCount: 42,
    });
    mockJson("dependabotalerts.json", { report: "dependabotalerts.json" });
    mockJson("http.json", { report: "http.json" });
    mockJson("lhr.json", { report: "lhr.json" });
    mockJson("nmapvuln.json", { report: "nmap.json" });
    mockJson("nuclei.json", [{ report: "nuclei.json" }]);
    mockJson("testssl.json", [{ report: "testssl.json" }]);
    mockJson("thirdparties.json", { report: "thirdparties.json" });
    mockJson("updownio.json", { report: "updownio.json" });
    mockJson("wappalyzer.json", { report: "wappalyzer.json" });
    mockJson("zap.json", { report: "zap.json" });
    mockJson("stats.json", { report: "stats.json" });
    mockJson("budget_page.json", { report: "budget_page.json" });
    mockJson("github_repository.json", { report: "github_repository.json" });
    mockJson("404.json", { broken: [1, 2, 3] });
    mockJson("declaration-a11y.json", {
      mention: "AccessibilitÃ© : partiellement conforme",
    });
    mockJson("declaration-rgpd.json", [
      {
        slug: "ml",
        declarationUrl: "https://declaration-ml.test",
        maxScore: 4,
        score: 3,
      },
      {
        slug: "pc",
        declarationUrl: "https://declaration-pc.test",
        maxScore: 4,
        score: 3,
      },
    ]);

    expect(
      generateUrlReport({
        url: testUrl,
      })
    ).toMatchSnapshot();

    // unmock all required files
    unMockJson("codescanalerts.json");
    unMockJson("dependabotalerts.json");
    unMockJson("http.json");
    unMockJson("lhr.json");
    unMockJson("nmapvuln.json");
    unMockJson("nuclei.json", []);
    unMockJson("testssl.json");
    unMockJson("thirdparties.json");
    unMockJson("updownio.json");
    unMockJson("wappalyzer.json");
    unMockJson("zap.json");
    unMockJson("stats.json");
    unMockJson("budget_page.json");
    unMockJson("github_repository.json");
    unMockJson("404.json");
    unMockJson("declaration-a11y.json");
    unMockJson("declaration-rgpd.json");
  });
  test(`should allow empty/invalid reports`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => false); // screenshot

    mockJson("codescanalerts.json");
    mockJson("wappalyzer.json");

    expect(
      generateUrlReport({
        url: testUrl,
      })
    ).toMatchSnapshot();

    unMockJson("codescanalerts.json");
    unMockJson("wappalyzer.json");
  });

  test(`should detect screenshot if any`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => true); // screenshot
    expect(
      generateUrlReport({
        url: testUrl,
      }).screenshot
    ).toEqual(true);
  });
});
