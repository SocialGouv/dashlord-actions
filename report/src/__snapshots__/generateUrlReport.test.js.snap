// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateUrlReport should allow empty/invalid reports 1`] = `
Object {
  "404": null,
  "betagouv": null,
  "budget_page": null,
  "codescan": Object {
    "report": "codescanalerts.json",
    "totalCount": 42,
  },
  "declaration-a11y": null,
  "declaration-rgpd": null,
  "dependabot": null,
  "ecoindex": null,
  "github_repository": null,
  "http": null,
  "lhr": null,
  "nmap": null,
  "nuclei": Array [],
  "screenshot": false,
  "sonarcloud": null,
  "stats": null,
  "summary": Object {
    "codescanCount": 42,
    "codescanGrade": undefined,
    "declaration-a11y": undefined,
    "declaration-rgpd": undefined,
  },
  "testssl": null,
  "thirdparties": null,
  "updownio": null,
  "url": "https://www.test.com",
  "wappalyzer": Object {
    "report": "wappalyzer.json",
  },
  "zap": null,
}
`;

exports[`generateUrlReport should generate latest report for a valid url 1`] = `
Object {
  "404": Array [
    1,
    2,
    3,
  ],
  "betagouv": null,
  "budget_page": Object {
    "report": "budget_page.json",
  },
  "codescan": Object {
    "report": "codescanalerts.json",
    "totalCount": 42,
  },
  "declaration-a11y": Object {
    "mention": "Accessibilit√© : partiellement conforme",
  },
  "declaration-rgpd": Array [
    Object {
      "declarationUrl": "https://declaration-ml.test",
      "maxScore": 4,
      "score": 3,
      "slug": "ml",
    },
    Object {
      "declarationUrl": "https://declaration-pc.test",
      "maxScore": 4,
      "score": 3,
      "slug": "pc",
    },
  ],
  "dependabot": Object {
    "report": "dependabotalerts.json",
  },
  "ecoindex": null,
  "github_repository": Object {
    "report": "github_repository.json",
  },
  "http": Object {
    "report": "http.json",
  },
  "lhr": Array [
    Object {
      "audits": Object {},
      "categories": Object {},
      "fetchTime": undefined,
      "finalUrl": undefined,
      "requestedUrl": undefined,
      "runWarnings": undefined,
    },
  ],
  "nmap": Object {
    "report": "nmap.json",
  },
  "nuclei": Array [],
  "screenshot": false,
  "sonarcloud": Array [
    Object {
      "repo": "sensgithub/eHospital",
      "result": Object {
        "analysisDate": "2023-04-03T01:37:49+0200",
        "commit": Object {
          "date": "2023-04-03T01:37:44+0200",
          "message": "Delete codeql.yml",
          "sha": "28f2cf43a343f5215ffa1052d9c659333626a2a4",
        },
        "isMain": true,
        "name": "main",
        "status": Object {
          "bugs": 174,
          "codeSmells": 2198,
          "vulnerabilities": 23,
        },
        "type": "LONG",
      },
    },
    Object {
      "repo": "zabbix/zabbix",
      "result": Object {
        "analysisDate": "2023-04-02T02:17:15+0200",
        "commit": Object {
          "date": "2023-03-31T21:45:03+0200",
          "message": ".......... [ZBX-1357] automatic update of translation strings",
          "sha": "f9d2062ed6560094ad6f4c8772cd82b72328cd83",
        },
        "isMain": true,
        "name": "master",
        "status": Object {
          "bugs": 73,
          "codeSmells": 13669,
          "qualityGateStatus": "ERROR",
          "vulnerabilities": 26,
        },
        "type": "LONG",
      },
    },
  ],
  "stats": Object {
    "report": "stats.json",
  },
  "summary": Object {
    "404": 3,
    "codescanCount": 42,
    "codescanGrade": undefined,
    "declaration-a11y": "F",
    "declaration-rgpd-ml": "D",
    "declaration-rgpd-pc": "D",
    "sonarcloudGrade": "F",
  },
  "testssl": Array [
    Object {
      "report": "testssl.json",
    },
  ],
  "thirdparties": Object {
    "report": "thirdparties.json",
  },
  "updownio": Object {
    "report": "updownio.json",
  },
  "url": "https://www.test.com",
  "wappalyzer": Object {
    "report": "wappalyzer.json",
  },
  "zap": Object {
    "report": "zap.json",
    "site": undefined,
  },
}
`;
